python/packages/isce3/math/offsets_polyfit.py

```diff
@@ -0,0 +1,508 @@
+        A = build_design_matrix(Ln, Pn, degree)
+
+        # Weighted normal equations
+        A_til = A * W
+        At = A_til.T
+        Nmat = At @ A_til
+        rhsL, rhsP = At @ (yL * W), At @ (yP * W)
+
+        try:
+            xL, Lc = cholesky_solve(Nmat, rhsL)
+            xP, _  = cholesky_solve(Nmat, rhsP)
+        except np.linalg.LinAlgError:
+            I = np.eye(Nmat.shape[0])
+            xL, Lc = cholesky_solve(Nmat + eps * I, rhsL)
+            xP, _  = cholesky_solve(Nmat + eps * I, rhsP)
+
+        Qx_hat = invert_from_cholesky(Lc)
+
+        # Residuals
+        yL_hat, yP_hat = A @ xL, A @ xP
+        eL, eP = yL - yL_hat, yP - yP_hat
+
+        # diag_Qe = Qy_diag - Qyhat_diag with weights applied in Qy
+        Qyhat_diag = np.einsum("ij,jk,ik->i", A, Qx_hat, A)
+        diag_Qe = np.clip(Qy_diag - Qyhat_diag, eps, None)
+
+        # w-test
+        s = np.sqrt(diag_Qe)
+        wL, wP = (eL[:, 0] / (s * sigmaL)), (eP[:, 0] / (s * sigmaP))
+        max_any = max(np.abs(wL).max(), np.abs(wP).max())
+        if (max_any <= crit_value) or (iteration >= max_iterations):
+            return {
+                "coefL": xL[:, 0],
+                "coefP": xP[:, 0],
+                "inliers": data,
+                "removed_indices": removed_indices,
+                "degree": degree,
+                "design_nunk": Nunk,
+            }
+
+        # Remove worst outlier
+        worst = int(np.argmax(wL * wL + wP * wP))
+        removed_indices.append(int(data[worst, 0]))
+        data = np.delete(data, worst, axis=0)
+        iteration += 1
+
+def _poly_design(line, pixel, degree, minL, maxL, minP, maxP):
+    """
+    Build a single 2D polynomial design vector for given line and pixel coordinates.
+
+    The polynomial terms are ordered consistently with the convention used in
+    polynomial fitting: for each total degree ``j = 0..degree`` and partial degree
+    ``k = 0..j``, the term is defined as ``l^(j-k) * p^k``, where ``l`` and ``p``
+    are the normalized line and pixel coordinates within [0, 1].
+
+    Parameters
+    ----------
+    line : float
+        Line coordinate of the tie point.
+    pixel : float
+        Pixel coordinate of the tie point.
+    degree : int
+        Total degree of the 2D polynomial.
+    minL, maxL : float
+        Minimum and maximum line coordinates used for normalization.
+    minP, maxP : float
+        Minimum and maximum pixel coordinates used for normalization.
+
+    Returns
+    -------
+    a : (M,) ndarray
+        Design vector containing all polynomial terms up to the given degree,
+        where ``M = (degree + 1) * (degree + 2) // 2``.
+    """
+
+    l = 0.0 if maxL == minL else (line - minL) / (maxL - minL)
+    p = 0.0 if maxP == minP else (pixel - minP) / (maxP - minP)
+
+    row = []
+    for j in range(degree + 1):
+        for k in range(j + 1):
+            row.append((l ** (j - k)) * (p ** k))
+    return np.array(row, dtype=float)
```

Comment:
```suggestion
    return build_design_matrix(np.array(l), np.array(p), degree)[0]
```
Seems better not to duplicate this fiddly stuff.

Reply:
this function has been deleted